web/packages/teleport/src/Assist/service.ts

```diff
@@ -0,0 +1,228 @@
+    id,
+    title: title || 'New Conversation',
+    created: new Date(created_time),
+  }));
+}
+
+export async function resolveServerMessage(
+  message: ServerMessage,
+  clusterId: string
+): Promise<ResolvedServerMessage> {
+  switch (message.type) {
+    case ServerMessageType.Command:
+      return resolveServerCommandMessage(message);
+
+    case ServerMessageType.CommandResult:
+      return resolveServerCommandResultMessage(message, clusterId);
+
+    case ServerMessageType.Assist:
+    case ServerMessageType.User:
+      return {
+        type: message.type,
+        message: message.payload,
+        created: new Date(message.created_time),
+      };
+  }
+}
+
+export async function getSessionEvents(sessionUrl: string): Promise<{
+  events: SessionEvent[];
+}> {
+  const response = await api.fetch(sessionUrl + '/events');
+
+  if (response.status !== 200) {
+    throw new Error('No session recording. The command execution failed.');
+  }
+
+  return response.json();
+}
+
+export async function getSessionStream(sessionUrl: string) {
+  const stream = await api.fetch(sessionUrl + '/stream?offset=0&bytes=4096', {
+    Accept: 'text/plain',
+    'Content-Type': 'text/plain; charset=utf-8',
+  });
+
+  if (stream.status === 200) {
+    return stream.text();
+  }
+}
+
+function isExecEvent(e: SessionEvent): e is ExecEvent {
+  return e.event == EventType.EXEC;
+}
+
+export function getNodesFromQuery(query: string, clusterId: string) {
+  const ns = new NodeService();
+
+  return ns.fetchNodes(clusterId, {
+    query,
+    limit: 100, // TODO: What if there is more nodes?
+  });
+}
+
+export async function getLoginsForQuery(query: string, clusterId: string) {
+  const nodes = await getNodesFromQuery(query, clusterId);
+
+  if (!nodes.agents.length) {
+    throw new Error('No nodes match the query');
+  }
+
+  const availableLogins = findIntersection(nodes.agents.map(e => e.sshLogins));
+
+  if (!availableLogins.length) {
+    throw new Error('No available logins found');
+  }
+
+  return sortLoginsWithRootLoginsLast(availableLogins);
+}
+
+export async function resolveServerCommandResultMessage(
+  message: ServerMessage,
+  clusterId: string
+): Promise<ResolvedCommandResultServerMessage> {
+  const payload = JSON.parse(message.payload) as CommandResultPayload;
+
+  try {
+    const sessionUrl = cfg.getSshPlaybackPrefixUrl({
+      clusterId,
+      sid: payload.session_id,
+    });
+
+    let output = await getSessionStream(sessionUrl);
+
+    if (!output) {
+      const events = await getSessionEvents(sessionUrl);
+      const execEvent = events.events?.find(isExecEvent);
+
+      output = execEvent?.exitError || 'Empty output';
+    }
+
+    return {
+      type: ServerMessageType.CommandResult,
+      nodeId: payload.node_id,
+      nodeName: payload.node_name,
+      executionId: payload.execution_id,
+      sessionId: payload.session_id,
+      created: new Date(message.created_time),
+      output,
+    };
+  } catch (err) {
```

Comment:
Hm, I skipped it because we aren't throwing any errors that aren't a Error or an extension of Error (NetworkError, etc) in the code that's called in the try